### ISSUE TITLE
[질문] Lab 05 - CSV reading using tf.decode_csv

### ISSUE THREAD

**Issue by socome** [moduPlayGround/DeepLearning-Basic#15]
lab 05에 Exercise에서 나온 'CSV reading using tf.decode_csv' 방법 해보신 분 계신가요?

[https://github.com/socome/ML_study/blob/master/lec05_2_exercise.ipynb](url)

노력해보고 있는데 쉽게 접근하기가 어려워서 이미 해보신 분들의 조언을 구하고자 합니다.

**Comment by lini1634** [comment]
# cost/loss function
cost = -tf.reduce_mean(Y*tf.log(hypothesis)+(1-Y)*tf.log(1-hypothesis)) 이렇게 되어있는데 
reduce_sum 이 빠진것같아요
cost=tf.reduce_mean(-tf.reduce_sum(Y*tf.log(hypothesis)+(1-Y)*tf.log(1-hypothesis))로 해보세욥

**Comment by socome** [comment]
https://github.com/socome/ML_study/blob/master/Lec05-2(realdatatest).ipynb

단순히 이 코드에서 batch만 추가해보려고 조작해서 이후 강의에서 나오는 reduce_sum을 사용할 
방법을 생각해보지 않았습니다. 추가하고 다시 실행시켜봐야겠네요. 
추가로 구한 W,b를 통한 정확도를 확인하려면 어떻게 코딩할 수 있을까요?

**Comment by lini1634** [comment]
reduce_sum이 문제가 아닌것같네요 ㅠ 이전 강의에서 cost function을 
tf.reduce_mean(-tf.reduce_sum(Y*tf.log(hypothesis)+(1-Y)*tf.log(1-hypothesis)))
라고 강의자료에 나와있었는데 구현할때는 또 tf.reduce_sum이 없네요
더 공부해봐야겠습니당 ..

**Comment by unizard** [comment]
본 이슈를 해결해주시는 분! 스벅 쏩니다. 두둥!

**Comment by socome** [comment]
https://www.google.com/url?sa=i&source=web&cd=&ved=2ahUKEwjUr-SFk-PeAhUHfd4KHanABEEQzPwBegQIARAC&url=https%3A%2F%2Fde-novo.org%2F2018%2F05%2F28%2Fbatch-normalization-%25EC%259D%25B4%25ED%2595%25B4%25ED%2595%2598%25EA%25B8%25B0%2F&psig=AOvVaw025acHHOPqGuJWmn6EzApH&ust=1542809405407225

해당질문관련 자료를 찾은것 같습니다.

**Comment by lini1634** [comment]
https://github.com/socome/ML_study/blob/master/Lec05-2(realdatatest).ipynb

단순히 batch만 사용하려고합니다.
추가로 구한 W,b값의 정확도를 확인하려면 어떻게 코딩할 수 있을까요?

socome님이 올려주신 저 코드에 학습률 높이고 더 많이 출력하니까 
0 0.9638013
500 0.47303903
1000 0.47186226
1500 0.47170252
2000 0.47167614
2500 0.47167164
3000 0.4716709
3500 0.47167075
4000 0.4716708
4500 0.4716708
5000 0.47167072
5500 0.47167072
6000 0.47167072
6500 0.47167072
7000 0.47167072
7500 0.47167072
8000 0.47167072
8500 0.47167072
9000 0.47167072
9500 0.47167072
10000 0.47167072
10500 0.47167072
11000 0.47167072
11500 0.47167072
12000 0.47167072
12500 0.47167072
13000 0.47167072
13500 0.47167072
14000 0.47167072
14500 0.47167072
15000 0.47167072
15500 0.47167072
16000 0.47167072
16500 0.47167072
17000 0.47167072
17500 0.47167072
18000 0.47167072
18500 0.47167072
19000 0.47167072
19500 0.47167072
20000 0.47167072

cost가 엄청 작진 않지만 일정한 값으로 나오는데 이게 그럼 최소화 된거 아닌가요??
local minimize 인건가요..?/

**Comment by socome** [comment]
일단 위의 코드는 일반 예제파일 코드인것 같습니다. 
(batch 코드가 추가되지 않은 예제 파일입니다.)

이후 강의를 들으면서 어떻게 해결한것 같기도 하면서 아닌것 같기도 한데..
https://github.com/socome/ML_study/blob/master/Lec05_2_exercise_final.ipynb

Epoch와 batch를 사용해서 ('lab 7-2'에서 일반적이라고 하네요..) 하니까 avg_cost가 감소하였고,
7-2에서 배운것처럼 diabetes.csv 파일의 일부는 train하는데, 일부는 test하는데 사용했을때 
정확도가 70퍼센트정도 나왔습니다. 

다만!!!!!!!!! 여기서 또 궁금증이 생긴것은 기존에는 단순히 sess을 돌리고 cost를 출력했다면
이번에는 avg_cost를 출력했는데, avg_cost와 실제 cost_val의 차이가 무엇인지 모르겠습니다.

**Comment by lini1634** [comment]
socome님이 올려주신 저 코드에서 
from google.colab import files
uploaded = files.upload() 
을  추가해서 제가 미리 다운 받아놓은 data-03-diabetes.csv파일 사용했어요
data-03-diabetes.csv(application/vnd.ms-excel) - 53438 bytes, last modified: 2018. 11. 19. - 100% done
 Saving data-03-diabetes.csv to data-03-diabetes.csv
그래서 batch코드 추가한 파일 맞습니당 ~
아까 socome님이 올려주신 자료 보면서 더 공부해보려 했는데 자료 지우셨네요 ㅠ 
혹시 다시 올려주실수 있을까요 ..?

**Comment by socome** [comment]
어떤자료를 올려드리면될까요?!

batch normalization에 관한 자료들 이야기
하시는거 갇아서 일단 링크 공유해드릴게요!!

http://sacko.tistory.com/m/44

**Comment by hyj378** [comment]
avg_cost의 cost_val차이가 궁금해서 올려주신 코드 뒷부분만 조금 바꿔봤는데

```     for epoch in range(training_epochs) :
      avg_cost = 0
      total_batch = int(524/batch_size_a)
      for i in range(total_batch):
          x_batch, y_batch = sess.run([train_x_batch, train_y_batch])
          cost_val, _ = sess.run([cost,train], feed_dict={X:x_batch, Y:y_batch})  
          print("\nchange :", avg_cost)
          avg_cost += cost_val / total_batch
      
      print("\n\n\nCost:",avg_cost) 
            
    coord.request_stop()
    coord.join(threads)
 ```
cost는 batch데이터 덩어리를 기준으로 구한 오차값이고
avg_cost는 덩어리를 데이터로 쪼개 구한 오차값들을 평균낸 것 인것 같은데
구분구적법으로 적분을 배우면서 등분 수를 늘리면 넓이가 더 정확해진다고 배웠던것 처럼
avg_cost를 통해 **더 정확한 오차값**을 얻을 수 있는것 아닐까 생각해봤습니다...
